File: apiEZPZ/src/main/java/com/example/apiezpz/post/controller/PostController.java
package com.example.ez.post.controller;

import com.example.ez.post.domain.Post;
import com.example.ez.post.service.Service;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/api/posts")
@CrossOrigin(origins = "http://localhost:3000") // React 연결 허용

public class PostController {
    private final Service service;
    private static final Logger log = LoggerFactory.getLogger(PostController.class);

    public PostController(Service Service) {
        this.service = Service;
    }

    @GetMapping
    public List<Post> getAllPosts() {
        List<Post> posts = service.getAllPosts();
        log.info("Fetched posts: {}", posts);
        return posts;
    }

    @GetMapping("/{id}")
    public ResponseEntity<Post> getPostById(@PathVariable Long id) {
        Optional<Post> post = service.getPostById(id);
        return post.map(ResponseEntity::ok).orElse(ResponseEntity.notFound().build());
    }

    @PostMapping
    public Post createPost(@RequestBody Post post) {
        return service.savePost(post);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Post> updatePost(@PathVariable Long id, @RequestBody Post postDetails) {
        Post post = service.getPostById(id)
                .orElseThrow(() -> new RuntimeException("게시글을 찾을 수 없습니다."));
                
        // 작성자 검증
        if (!post.getWriter().equals(postDetails.getWriter())) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
        
        return ResponseEntity.ok(service.updatePost(id, postDetails));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deletePost(@PathVariable Long id, @RequestParam String writer) {
        Post post = service.getPostById(id)
                .orElseThrow(() -> new RuntimeException("게시글을 찾을 수 없습니다."));
                
        // 작성자 검증
        if (!post.getWriter().equals(writer)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
        
        service.deletePost(id);
        return ResponseEntity.ok().build();
    }
}

File: apiEZPZ/src/main/kotlin/com/example/apiezpz/post/controller/PostController.kt
package com.example.apiezpz.post.controller

import com.example.apiezpz.post.domain.Post
import com.example.apiezpz.post.service.PostService
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/posts")
class PostController(
    private val postService: PostService
) {
    private val log = LoggerFactory.getLogger(PostController::class.java)

    // 모든 게시글 조회 (인증 불필요)
    @GetMapping
    fun getAllPosts(): ResponseEntity<List<Post>> {
        log.info("모든 게시글 조회 요청")
        val posts = postService.getAllPosts()
        log.info("게시글 조회 완료 - 조회된 게시글 수: {}", posts.size)
        return ResponseEntity.ok(posts)
    }

    // 특정 게시글 조회 (인증 불필요)
    @GetMapping("/{id}")
    fun getPostById(@PathVariable id: Long): ResponseEntity<Post> {
        log.info("게시글 조회 요청 - 게시글ID: {}", id)
        val post = postService.getPostById(id)
        return if (post.isPresent) {
            log.info("게시글 조회 성공 - 게시글ID: {}", id)
            ResponseEntity.ok(post.get())
        } else {
            log.warn("게시글을 찾을 수 없음 - 게시글ID: {}", id)
            ResponseEntity.notFound().build()
        }
    }

    // 게시글 작성 (인증 필요)
    @PostMapping
    @PreAuthorize("isAuthenticated()")
    fun createPost(@RequestBody post: Post, 
                   @AuthenticationPrincipal userDetails: UserDetails): ResponseEntity<Post> {
        log.info("게시글 작성 요청 - 사용자: {}, 제목: {}", userDetails.username, post.title)
        
        // 작성자 정보 설정
        post.writer = userDetails.username
        
        val savedPost = postService.savePost(post)
        log.info("게시글 작성 완료 - 사용자: {}, 게시글ID: {}", userDetails.username, savedPost.id)
        return ResponseEntity.ok(savedPost)
    }

    // 게시글 수정 (인증 필요)
    @PutMapping("/{id}")
    @PreAuthorize("isAuthenticated()")
    fun updatePost(@PathVariable id: Long, 
                   @RequestBody postDetails: Post,
                   @AuthenticationPrincipal userDetails: UserDetails): ResponseEntity<Post> {
        log.info("게시글 수정 요청 - 사용자: {}, 게시글ID: {}", userDetails.username, id)
        
        return try {
            val updatedPost = postService.updatePost(id, postDetails)
            log.info("게시글 수정 완료 - 사용자: {}, 게시글ID: {}", userDetails.username, id)
            ResponseEntity.ok(updatedPost)
        } catch (e: RuntimeException) {
            log.warn("게시글 수정 실패 - 사용자: {}, 게시글ID: {}, 오류: {}", userDetails.username, id, e.message)
            ResponseEntity.notFound().build()
        }
    }

    // 게시글 삭제 (인증 필요)
    @DeleteMapping("/{id}")
    @PreAuthorize("isAuthenticated()")
    fun deletePost(@PathVariable id: Long,
                   @AuthenticationPrincipal userDetails: UserDetails): ResponseEntity<Void> {
        log.info("게시글 삭제 요청 - 사용자: {}, 게시글ID: {}", userDetails.username, id)
        
        return try {
            postService.deletePost(id)
            log.info("게시글 삭제 완료 - 사용자: {}, 게시글ID: {}", userDetails.username, id)
            ResponseEntity.ok().build()
        } catch (e: Exception) {
            log.warn("게시글 삭제 실패 - 사용자: {}, 게시글ID: {}, 오류: {}", userDetails.username, id, e.message)
            ResponseEntity.notFound().build()
        }
    }
} 